File: ios/LaneShadow/Views/Molecules/NewSessionButton.swift
import LaneShadowTheme
import SwiftUI

// MARK: - New Session Button Variant Enum

/**
 * New session button variant
 *
 * Following RN wrapper API from react-native/components/ui/new-session-button.tsx
 */
public enum LSNewSessionVariant {
    case header
    case fab
    case text
}

// MARK: - New Session Button Size Enum

/**
 * New session button size
 *
 * Following RN wrapper API from react-native/components/ui/new-session-button.tsx
 */
public enum LSNewSessionSize {
    case sm
    case md
    case lg
}

// MARK: - New Session Button Component

/**
 * New session button component
 *
 * Following RN wrapper API from react-native/components/ui/new-session-button.tsx
 * Matrix reference: .spec/prds/native-rewrite/matrices/ui/molecules/NewSessionButton.md
 *
 * ## Design Tokens Used
 * - Colors: `theme.colors.primary`, `theme.colors.onSurface`, `theme.colors.surfaceVariant`
 * - Typography: `theme.type.body` with weight 600/700
 * - Spacing: `theme.space.sm/md` (6-8pt for gap)
 * - Elevation: `theme.elevation.level3` for fab variant
 * - Opacity: 0.5 when disabled, 0.8 when pressed
 *
 * ## Variants
 * - **header** (default): HStack with plus.circle icon + label, muted text color, 6pt spacing
 * - **fab**: Circular floating action button with elevation, 50% opacity when disabled
 * - **text**: HStack with plus.circle icon + label, primary color, 8pt spacing, bold text
 *
 * ## Size configurations
 * - **sm**: icon 20pt, font 13pt, padding 4pt
 * - **md**: icon 24pt, font 14pt, padding 6pt (default)
 * - **lg**: icon 28pt, font 16pt, padding 8pt
 *
 * ## Parameters
 * - variant: Button variant (header, fab, text)
 * - label: Button label text (default: "Session")
 * - size: Button size (sm, md, lg)
 * - onPress: Action callback when button is pressed
 * - disabled: Whether button is disabled (default: false)
 * - accessibilityLabel: Accessibility label for screen readers
 * - testID: Test identifier for UI testing
 */
public struct LSNewSessionButton: View {
    @Environment(\.theme) private var theme
    @State private var isPressed = false

    private let variant: LSNewSessionVariant
    private let label: String
    private let size: LSNewSessionSize
    private let onPress: (() -> Void)?
    private let disabled: Bool
    private let accessibilityLabel: String?
    private let testID: String?

    public init(
        variant: LSNewSessionVariant = .header,
        label: String = "Session",
        size: LSNewSessionSize = .md,
        onPress: (() -> Void)? = nil,
        disabled: Bool = false,
        accessibilityLabel: String? = nil,
        testID: String? = nil
    ) {
        self.variant = variant
        self.label = label
        self.size = size
        self.onPress = onPress
        self.disabled = disabled
        self.accessibilityLabel = accessibilityLabel
        self.testID = testID
    }

    // MARK: - Layout Computed Properties

    private var iconSize: CGFloat {
        switch size {
        case .sm:
            return 20
        case .lg:
            return 28
        case .md:
            return 24
        }
    }

    private var fontSize: CGFloat {
        switch size {
        case .sm:
            return 13
        case .lg:
            return 16
        case .md:
            return 14
        }
    }

    private var padding: CGFloat {
        switch size {
        case .sm:
            return theme.space.xs // 4
        case .lg:
            return theme.space.md // 8
        case .md:
            return theme.space.sm // 6
        }
    }

    private var fontWeight: Font.Weight {
        switch variant {
        case .text:
            return .bold // 700
        case .header, .fab:
            return .semibold // 600
        }
    }

    private var spacing: CGFloat {
        switch variant {
        case .header:
            return 6
        case .fab:
            return 0
        case .text:
            return 8
        }
    }

    // MARK: - Color Computed Properties

    private var iconColor: Color {
        if disabled {
            return theme.colors.onSurface.subtle
        }

        switch variant {
        case .header:
            return theme.colors.primary.default
        case .fab:
            return theme.colors.onPrimary.default
        case .text:
            return theme.colors.primary.default
        }
    }

    private var textColor: Color {
        if disabled {
            return theme.colors.onSurface.subtle
        }

        switch variant {
        case .header:
            return theme.colors.onSurface.muted
        case .fab:
            return theme.colors.onPrimary.default
        case .text:
            return theme.colors.primary.default
        }
    }

    private var backgroundColor: Color {
        if disabled {
            return theme.colors.surfaceVariant.default
        }

        if isPressed {
            return theme.colors.primary.pressed ?? theme.colors.primary.default
        }

        return theme.colors.primary.default
    }

    // MARK: - Body

    public var body: some View {
        Group {
            switch variant {
            case .fab:
                fabContent
            case .header:
                headerContent
            case .text:
                textContent
            }
        }
        .disabled(disabled)
        .opacity(disabled ? 0.5 : (isPressed ? 0.8 : 1.0))
        .accessibilityAddTraits(.isButton)
        .accessibilityAddTraits(disabled ? .notEnabled : [])
        .accessibilityLabel(accessibilityLabel ?? computedAccessibilityLabel)
        .accessibilityIdentifier(testID ?? "new-session-button")
    }

    // MARK: - FAB Content

    private var fabContent: some View {
        let fabSize: CGFloat = {
            switch size {
            case .sm:
                return 48
            case .lg:
                return 64
            case .md:
                return 56
            }
        }()

        return Button(action: {
            onPress?()
        }) {
            LSIconSymbol(
                name: "plus",
                size: iconSize,
                color: iconColor
            )
            .frame(width: fabSize, height: fabSize)
            .background(backgroundColor)
            .clipShape(Circle())
            .shadow(
                color: theme.elevation.level3.shadowColor,
                radius: theme.elevation.level3.radius,
                x: theme.elevation.level3.offsetX,
                y: theme.elevation.level3.offsetY
            )
        }
        .buttonStyle(.plain)
        .simultaneousGesture(
            DragGesture(minimumDistance: 0)
                .onChanged { _ in
                    if !disabled {
                        isPressed = true
                    }
                }
                .onEnded { _ in
                    isPressed = false
                }
        )
    }

    // MARK: - Header Content

    private var headerContent: some View {
        Button(action: {
            onPress?()
        }) {
            HStack(spacing: spacing) {
                LSIconSymbol(
                    name: "plus-circle-outline",
                    size: iconSize,
                    color: iconColor
                )

                Text(label)
                    .font(.system(size: fontSize, weight: fontWeight))
                    .foregroundStyle(textColor)
            }
            .padding(padding)
        }
        .buttonStyle(.plain)
        .simultaneousGesture(
            DragGesture(minimumDistance: 0)
                .onChanged { _ in
                    if !disabled {
                        isPressed = true
                    }
                }
                .onEnded { _ in
                    isPressed = false
                }
        )
    }

    // MARK: - Text Content

    private var textContent: some View {
        Button(action: {
            onPress?()
        }) {
            HStack(spacing: spacing) {
                LSIconSymbol(
                    name: "plus-circle-outline",
                    size: iconSize,
                    color: iconColor
                )

                Text(label)
                    .font(.system(size: fontSize, weight: fontWeight))
                    .foregroundStyle(textColor)
            }
            .padding(padding)
        }
        .buttonStyle(.plain)
        .simultaneousGesture(
            DragGesture(minimumDistance: 0)
                .onChanged { _ in
                    if !disabled {
                        isPressed = true
                    }
                }
                .onEnded { _ in
                    isPressed = false
                }
        )
    }

    // MARK: - Helpers

    private var computedAccessibilityLabel: String {
        switch variant {
        case .fab:
            return "New session"
        case .header, .text:
            return "New \(label)"
        }
    }
}

// MARK: - Preview

#Preview("NewSessionButton - Header") {
    VStack(spacing: 16) {
        LSNewSessionButton(
            variant: .header,
            label: "Session",
            size: .sm,
            onPress: {}
        )

        LSNewSessionButton(
            variant: .header,
            label: "Session",
            size: .md,
            onPress: {}
        )

        LSNewSessionButton(
            variant: .header,
            label: "Session",
            size: .lg,
            onPress: {}
        )

        LSNewSessionButton(
            variant: .header,
            label: "Session",
            size: .md,
            onPress: {},
            disabled: true
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("NewSessionButton - FAB") {
    VStack(spacing: 16) {
        LSNewSessionButton(
            variant: .fab,
            label: "Session",
            size: .sm,
            onPress: {}
        )

        LSNewSessionButton(
            variant: .fab,
            label: "Session",
            size: .md,
            onPress: {}
        )

        LSNewSessionButton(
            variant: .fab,
            label: "Session",
            size: .lg,
            onPress: {}
        )

        LSNewSessionButton(
            variant: .fab,
            label: "Session",
            size: .md,
            onPress: {},
            disabled: true
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("NewSessionButton - Text") {
    VStack(spacing: 16) {
        LSNewSessionButton(
            variant: .text,
            label: "Session",
            size: .sm,
            onPress: {}
        )

        LSNewSessionButton(
            variant: .text,
            label: "Session",
            size: .md,
            onPress: {}
        )

        LSNewSessionButton(
            variant: .text,
            label: "Session",
            size: .lg,
            onPress: {}
        )

        LSNewSessionButton(
            variant: .text,
            label: "Session",
            size: .md,
            onPress: {},
            disabled: true
        )
    }
    .padding()
    .laneShadowTheme()
}

File: android/app/src/main/java/com/laneshadow/ui/components/molecules/NewSessionButton.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Create
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * NewSessionButton variant types
 *
 * Following RN wrapper API from react-native/components/ui/new-session-button.tsx
 */
enum class NewSessionVariant {
    Header,
    Fab,
    Text,
}

/**
 * NewSessionButton size variants
 *
 * Following RN wrapper API from react-native/components/ui/new-session-button.tsx
 */
enum class NewSessionSize {
    Sm,
    Md,
    Lg,
}

/**
 * NewSessionButton molecule component
 *
 * Button with 3 variants for creating new chat sessions.
 * Following React Native wrapper patterns from react-native/components/ui/new-session-button.tsx
 *
 * ## Usage
 *
 * ```kotlin
 * // Header variant (default)
 * NewSessionButton(
 *     onPress = { /* handle press */ },
 *     variant = NewSessionVariant.Header,
 *     label = "Session",
 *     size = NewSessionSize.Md,
 * )
 *
 * // FAB variant
 * NewSessionButton(
 *     onPress = { /* handle press */ },
 *     variant = NewSessionVariant.Fab,
 *     size = NewSessionSize.Md,
 * )
 *
 * // Text variant
 * NewSessionButton(
 *     onPress = { /* handle press */ },
 *     variant = NewSessionVariant.Text,
 *     label = "New Session",
 *     size = NewSessionSize.Md,
 * )
 * ```
 *
 * @param onPress Callback when button is pressed (null makes button non-interactive)
 * @param disabled Whether button is disabled (adds opacity and prevents interaction)
 * @param variant Button variant (header, fab, text)
 * @param label Optional text label for header and text variants (default: "Session")
 * @param size Button size variant (sm, md, lg) - affects icon size, font size, and padding
 * @param modifier Modifier for the button container
 * @param testID Test ID for UI testing
 */
@Composable
fun NewSessionButton(
    onPress: (() -> Unit)? = null,
    disabled: Boolean = false,
    variant: NewSessionVariant = NewSessionVariant.Header,
    label: String = "Session",
    size: NewSessionSize = NewSessionSize.Md,
    modifier: Modifier = Modifier,
    testID: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // Get size configuration
    val sizeConfig = getSizeConfig(size)

    // Track pressed state
    val interactionSource = remember { MutableInteractionSource() }
    val isPressed by interactionSource.collectIsPressedAsState()

    // Calculate opacity
    val opacity = when {
        disabled -> 0.5f
        isPressed -> 0.8f
        else -> 1f
    }

    // Determine if button is disabled (no onPress also disables)
    val isDisabled = disabled || onPress == null

    when (variant) {
        NewSessionVariant.Fab -> {
            FabVariant(
                onPress = onPress,
                disabled = isDisabled,
                isPressed = isPressed,
                size = size,
                sizeConfig = sizeConfig,
                opacity = opacity,
                modifier = modifier,
                testID = testID,
            )
        }
        NewSessionVariant.Text -> {
            TextVariant(
                onPress = onPress,
                disabled = isDisabled,
                isPressed = isPressed,
                label = label,
                sizeConfig = sizeConfig,
                opacity = opacity,
                modifier = modifier,
                testID = testID,
            )
        }
        NewSessionVariant.Header -> {
            HeaderVariant(
                onPress = onPress,
                disabled = isDisabled,
                isPressed = isPressed,
                label = label,
                sizeConfig = sizeConfig,
                opacity = opacity,
                modifier = modifier,
                testID = testID,
            )
        }
    }
}

/**
 * Size configuration data class
 *
 * Contains icon size, font size, and padding for each size variant
 */
private data class SizeConfig(
    val iconSize: Dp,
    val fontSize: Float,
    val padding: Dp,
)

/**
 * Get size configuration based on size variant
 *
 * Following RN wrapper behavior:
 * - sm: icon 20dp, font 13sp, padding 4dp
 * - md: icon 24dp, font 14sp, padding 6dp
 * - lg: icon 28dp, font 16sp, padding 8dp
 */
@Composable
private fun getSizeConfig(size: NewSessionSize): SizeConfig {
    val density = LocalDensity.current
    return when (size) {
        NewSessionSize.Sm -> SizeConfig(
            iconSize = with(density) { 20.dp },
            fontSize = 13f,
            padding = with(density) { 4.dp },
        )
        NewSessionSize.Lg -> SizeConfig(
            iconSize = with(density) { 28.dp },
            fontSize = 16f,
            padding = with(density) { 8.dp },
        )
        NewSessionSize.Md -> SizeConfig(
            iconSize = with(density) { 24.dp },
            fontSize = 14f,
            padding = with(density) { 6.dp },
        )
    }
}

/**
 * FAB variant implementation
 *
 * Circular floating action button with:
 * - Primary.default bg (pressed: primary.pressed, disabled: surfaceVariant.default)
 * - Plus icon in onPrimary.default (disabled: onSurface.subtle)
 * - Sizes: sm=48dp, md=56dp, lg=64dp (radius = size/2)
 * - Elevation: level 4
 */
@Composable
private fun FabVariant(
    onPress: (() -> Unit)?,
    disabled: Boolean,
    isPressed: Boolean,
    size: NewSessionSize,
    sizeConfig: SizeConfig,
    opacity: Float,
    modifier: Modifier = Modifier,
    testID: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // FAB size based on size variant
    val fabSize = when (size) {
        NewSessionSize.Sm -> 48.dp
        NewSessionSize.Lg -> 64.dp
        NewSessionSize.Md -> 56.dp
    }

    // Background color
    val backgroundColor = when {
        disabled -> theme.colors.surfaceVariant.default
        isPressed -> theme.colors.primary.pressed ?: theme.colors.primary.default
        else -> theme.colors.primary.default
    }

    // Icon color (use disabled color for subtle state)
    val iconColor = when {
        disabled -> theme.colors.onSurface.disabled ?: theme.colors.onSurface.default
        else -> theme.colors.onPrimary.default
    }

    // Elevation
    val elevation = theme.elevation.light.level4

    Surface(
        onClick = { onPress?.invoke() },
        modifier = modifier
            .size(fabSize)
            .alpha(opacity)
            .semantics {
                role = Role.Button
                contentDescription = "New session"
            },
        shape = CircleShape,
        color = backgroundColor,
        tonalElevation = elevation,
        shadowElevation = elevation,
        enabled = !disabled,
    ) {
        Icon(
            imageVector = Icons.Default.Add,
            contentDescription = "New session",
            tint = iconColor,
            modifier = Modifier.size(sizeConfig.iconSize),
        )
    }
}

/**
 * Text variant implementation
 *
 * Row with plus-circle-outline icon + label text:
 * - Icon: primary.default (disabled: onSurface.subtle)
 * - Text: primary.default, fontWeight 700
 * - Gap: 8dp
 */
@Composable
private fun TextVariant(
    onPress: (() -> Unit)?,
    disabled: Boolean,
    isPressed: Boolean,
    label: String,
    sizeConfig: SizeConfig,
    opacity: Float,
    modifier: Modifier = Modifier,
    testID: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // Icon color (use disabled color for subtle state)
    val iconColor = when {
        disabled -> theme.colors.onSurface.disabled ?: theme.colors.onSurface.default
        else -> theme.colors.primary.default
    }

    // Text color
    val textColor = when {
        disabled -> theme.colors.onSurface.disabled ?: theme.colors.onSurface.default
        else -> theme.colors.primary.default
    }

    Row(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        modifier = modifier
            .clickable(
                onClick = { onPress?.invoke() },
                enabled = !disabled,
                interactionSource = remember { MutableInteractionSource() },
                indication = null,
            )
            .alpha(opacity)
            .semantics {
                role = Role.Button
                contentDescription = "New $label"
            },
    ) {
            // Use Create icon as closest to "plus-circle-outline"
            Icon(
                imageVector = Icons.Default.Create,
                contentDescription = null,
                tint = iconColor,
                modifier = Modifier.size(sizeConfig.iconSize),
            )

            Text(
                text = label,
                style = androidx.compose.ui.text.TextStyle(
                    fontSize = with(LocalDensity.current) { sizeConfig.fontSize.toSp() },
                    fontWeight = FontWeight.Bold,
                ),
                color = textColor,
            )
    }
}

/**
 * Header variant implementation (default)
 *
 * Row with plus-circle-outline icon + label text:
 * - Icon: onSurface.subtle (disabled) or primary.default
 * - Text: onSurface.muted, fontWeight 600
 * - Gap: 6dp
 */
@Composable
private fun HeaderVariant(
    onPress: (() -> Unit)?,
    disabled: Boolean,
    isPressed: Boolean,
    label: String,
    sizeConfig: SizeConfig,
    opacity: Float,
    modifier: Modifier = Modifier,
    testID: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // Icon color (use disabled color for subtle state)
    val iconColor = when {
        disabled -> theme.colors.onSurface.disabled ?: theme.colors.onSurface.default
        else -> theme.colors.primary.default
    }

    // Text color - use default onSurface color (muted is not available in ColorSet)
    val textColor = theme.colors.onSurface.default

    Row(
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        modifier = modifier
            .clickable(
                onClick = { onPress?.invoke() },
                enabled = !disabled,
                interactionSource = remember { MutableInteractionSource() },
                indication = null,
            )
            .alpha(opacity)
            .semantics {
                role = Role.Button
                contentDescription = "New $label"
            },
    ) {
            // Use Create icon as closest to "plus-circle-outline"
            Icon(
                imageVector = Icons.Default.Create,
                contentDescription = null,
                tint = iconColor,
                modifier = Modifier.size(sizeConfig.iconSize),
            )

            Text(
                text = label,
                style = androidx.compose.ui.text.TextStyle(
                    fontSize = with(LocalDensity.current) { sizeConfig.fontSize.toSp() },
                    fontWeight = FontWeight.SemiBold,
                ),
                color = textColor,
            )
    }
}
